frontend/lib/src/components/widgets/Slider/Slider.tsx

```diff
@@ -282,159 +149,294 @@ class Slider extends React.PureComponent<Props, State> {
+              paddingRight: spacing.none,
+              paddingTop: spacing.twoThirdsSmFont,
             },
-            InnerTrack: {
-              style: ({ $disabled }: StyleProps) => ({
-                height: "4px",
-                ...($disabled ? { background: colors.darkenedBgMix25 } : {}),
-              }),
-            },
-            TickBar: this.renderTickBar,
-          }}
-        />
-      </div>
-    )
+          },
+          InnerTrack: {
+            style: innerTrackStyle,
+          },
+          TickBar: renderTickBar,
+        }}
+      />
+    </div>
+  )
+}
+
+function getStateFromWidgetMgr(
+  widgetMgr: WidgetStateManager,
+  element: SliderProto
+): number[] | undefined {
+  return widgetMgr.getDoubleArrayValue(element)
+}
+
+function getDefaultStateFromProto(element: SliderProto): number[] {
+  return element.default
+}
+
+function getCurrStateFromProto(element: SliderProto): number[] {
+  return element.value
+}
+
+function updateWidgetMgrState(
+  element: SliderProto,
+  widgetMgr: WidgetStateManager,
+  vws: ValueWithSource<number[]>,
+  fragmentId?: string
+): void {
+  widgetMgr.setDoubleArrayValue(
+    element,
+    vws.value,
+    { fromUi: vws.fromUi },
+    fragmentId
+  )
+}
+
+function isDateTimeType(element: SliderProto): boolean {
+  const { dataType } = element
+  return (
+    dataType === SliderProto.DataType.DATETIME ||
+    dataType === SliderProto.DataType.DATE ||
+    dataType === SliderProto.DataType.TIME
+  )
+}
+
+function formatValue(value: number, element: SliderProto): string {
+  const { format, options } = element
+  if (isDateTimeType(element)) {
+    // Python datetime uses microseconds, but JS & Moment uses milliseconds
+    // The timestamp is always set to the UTC timezone, even so, the actual timezone
+    // for this timestamp in the backend could be different.
+    // However, the frontend component does not need to know about the actual timezone.
+    return moment.utc(value / 1000).format(format)
+  }
+
+  if (options.length > 0) {
+    return sprintf(format, options[value])
+  }
+
+  return sprintf(format, value)
+}
+
+/**
+ * Return the value of the slider. This will either be an array with
+ * one value (for a single value slider), or an array with two
+ * values (for a range slider).
+ */
+function getValueAsArray(value: number[], element: SliderProto): number[] {
+  const { min, max } = element
+  let start = value[0]
+  let end = value.length > 1 ? value[1] : value[0]
+  // Adjust the value if it's out of bounds.
+  if (start > end) {
+    start = end
+  }
+  if (start < min) {
+    start = min
+  }
+  if (start > max) {
+    start = max
+  }
+  if (end < min) {
+    end = min
+  }
+  if (end > max) {
+    end = max
+  }
+  return value.length > 1 ? [start, end] : [start]
+}
+
+function alignValueOnThumb(
+  slider: HTMLDivElement | null,
+  thumb: HTMLDivElement | null,
+  thumbValue: HTMLDivElement | null
+): void {
+  if (slider && thumb && thumbValue) {
```

Comment:
super-nit (personal style): I personally prefer writing blocks that encapsulate many lines like this:

```typescript
function alignValueOnThumb(
  slider: HTMLDivElement | null,
  thumb: HTMLDivElement | null,
  thumbValue: HTMLDivElement | null
): void {
  if (!slider || !thumb || !thumbValue) {
    return
  } 

  const ...
} 
```

the reason is that we have one less level of nesting which makes code usually more pleasant and easier to read (its more so if the body itself contains additional nesting).

Reply:
Agreed! I was just trying not to modify the existing code too much, to avoid introducing bugs.